src/pacemaker/protocol.rs

```diff
@@ -127,46 +132,70 @@ impl<N: Network> Pacemaker<N> {
     /// Additionally, tick should check if there is a QC for the current view (whether an epoch-change 
     /// view or not) available in the block tree, and if so it should broadcast the QC in an advance 
     /// view message.
+    /// 
+    /// It should also check of the validator set state has been updated, and if so it should update the
+    /// timeout vote collectors accordingly.
     pub(crate) fn tick<K: KVStore>(&mut self, block_tree: &BlockTree<K>) -> Result<(), PacemakerError>{
 
         let cur_view = self.view_info.view;
+        let validator_set_state = block_tree.validator_set_state()?;
 
         // 1. Check if the current view has timed out, and proceed accordingly.
         if Instant::now() > self.view_info.deadline {
+
+            Event::ViewTimeout(ViewTimeoutEvent{timestamp: SystemTime::now(), view: cur_view}).publish(&self.event_publisher);
+
             if is_epoch_change_view(&cur_view, self.config.epoch_length) {
-                if block_tree.committed_validator_set().contains(&self.config.keypair.public()) {
+                if is_validator(&self.config.keypair.public(), &validator_set_state) {
                     let pacemaker_message = 
                         PacemakerMessage::timeout_vote(
                             &self.config.keypair, 
                             self.config.chain_id, 
                             cur_view, 
-                            block_tree.highest_tc()
+                            block_tree.highest_tc()?,
                         );
-                    self.sender.broadcast(Message::from(pacemaker_message));
+                    self.sender.broadcast(Message::from(pacemaker_message.clone()));
+                    if let PacemakerMessage::TimeoutVote(timeout_vote) = pacemaker_message {
+                        Event::TimeoutVote(TimeoutVoteEvent{timestamp: SystemTime::now(), timeout_vote}).publish(&self.event_publisher)
+                    }
                 }
                 self.extend_view()?
             } else {
-                self.update_view(cur_view + 1 , block_tree)?
+                self.update_view(cur_view + 1, &validator_set_state)?;
             }
+
+            return Ok(())
         }
 
-        // 2. Check if a QC for the current view is available and if I am a validator, and if so 
+        // 2. Check if the timeout vote collectors need to be updated in response to a validator set
+        //    state update. 
+        let _ = self.state.timeout_vote_collectors.update_validator_sets(&validator_set_state);
+        
+
+        // 3. Check if a QC for the current view is available and if I am a validator, and if so 
         //    broadcast an advance view message.
-        if block_tree.highest_qc().view == cur_view && block_tree.committed_validator_set().contains(&self.config.keypair.public()) {
+        if block_tree.highest_qc()?.view == cur_view 
```

Comment:
There are two problems with this `tick` function:
1. During block sync, a replica's block tree and highest QC could increase to much higher than the current view. After exiting block sync, the expected behavior is this scenario is for Pacemaker to jump to `highest_qc.view + 1`. However, in the current implementation the only way to jump to a higher view is to receive an `AdvanceView` message, and this check over here (`block_tree.highest_qc()?.view == cur_view`) prevents us from broadcasting an `AdvanceView` message. The fix is to replace `==` in the condition with `>=`.
2. Even if we make the first change, listeners will still *never* broadcast `AdvanceView` messages because of the condition in line 179. This means listeners are fully dependent on `AdvanceView` messages sent by validators to increase their local views. Which may be kind of odd, but is okay.